File: code/capture_gl.py
#!/usr/bin/env python3
# ============================================================
# capture_gl.py — カメラキャプチャ(独立スレッド) + OpenGL テクスチャ描画
#
# キャプチャ専用スレッドがカメラの上限速度でフレームを取り込み、
# メインスレッドは vsync(ディスプレイのリフレッシュレート)に合わせて
# 「最新フレームだけ」をテクスチャとして描画する。
# → 撮影レートが描画レートに律速されない。
#
# 使い方:
#   python capture_gl.py                # カメラ 0 を開く
#   python capture_gl.py --camera 1    # カメラ番号を指定
#   python capture_gl.py --test        # カメラ無しで合成映像を使う
#   python capture_gl.py --seconds 5   # 5秒後に自動終了(動作確認用)
#   ESC または ウィンドウを閉じると終了
#
# タイトルバーに camera fps(取り込み)と display fps(描画)を表示。
# camera fps が display fps を上回っていれば分離が効いている。
#
# 必要: pip install opencv-python glfw PyOpenGL
# ============================================================

import argparse
import threading
import time

import numpy as np
import cv2
import glfw
from OpenGL.GL import (
    GL_BGR, GL_CLAMP_TO_EDGE, GL_COLOR_BUFFER_BIT, GL_LINEAR, GL_QUADS,
    GL_RGB, GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_TEXTURE_MIN_FILTER,
    GL_TEXTURE_WRAP_S, GL_TEXTURE_WRAP_T, GL_UNPACK_ALIGNMENT,
    GL_UNSIGNED_BYTE, glBegin, glBindTexture, glClear, glClearColor,
    glDisable, glEnable, glEnd, glGenTextures, glPixelStorei, glTexCoord2f,
    glTexImage2D, glTexParameteri, glTexSubImage2D, glVertex2f, glViewport,
)


class CaptureThread:
    """カメラを専用スレッドで回し、常に最新フレームだけを保持する。

    cv2.VideoCapture.read() は C 側で GIL を解放するため、
    Python の threading でも取り込みと描画は実際に並列動作する。
    """

    def __init__(self, camera_id=0, test=False, width=1280, height=720, req_fps=120):
        self.test = test
        self.size = (width, height)
        self.lock = threading.Lock()
        self.frame = None          # 最新フレーム(BGR, np.ndarray)
        self.seq = 0               # フレーム通し番号(描画側の新着判定用)
        self.fps = 0.0             # 実測キャプチャfps(0.5秒窓)
        self.running = True

        if not test:
            self.cap = cv2.VideoCapture(camera_id)
            if not self.cap.isOpened():
                raise RuntimeError(
                    f"カメラ {camera_id} を開けません。--test で合成映像を試せます")
            # 高速化の要:MJPG圧縮・バッファ1枚・解像度/レートの要求
            self.cap.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter_fourcc(*"MJPG"))
            self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, width)
            self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
            self.cap.set(cv2.CAP_PROP_FPS, req_fps)
            self.cap.set(cv2.CAP_PROP_BUFFERSIZE, 1)

        self.thread = threading.Thread(target=self._loop, daemon=True)
        self.thread.start()

    def _make_test_frame(self, t):
        """カメラ無しでの動作確認用:動くグラデーションと時刻を描く"""
        w, h = self.size
        x = np.linspace(0, 1, w, dtype=np.float32)
        y = np.linspace(0, 1, h, dtype=np.float32)[:, None]
        phase = t * 2.0
        img = np.empty((h, w, 3), dtype=np.uint8)
        img[:, :, 0] = ((np.sin(x * 6.28 + phase) * 0.5 + 0.5) * 255)      # B
        img[:, :, 1] = ((np.cos(y * 6.28 + phase) * 0.5 + 0.5) * 255)      # G
        img[:, :, 2] = ((x * 0 + (np.sin(phase) * 0.5 + 0.5)) * 255)       # R
        cv2.putText(img, f"TEST {t:8.3f}s", (40, h // 2),
                    cv2.FONT_HERSHEY_SIMPLEX, 2.0, (255, 255, 255), 4)
        return img

    def _loop(self):
        n = 0
        win_t0 = time.perf_counter()
        t0 = win_t0
        while self.running:
            if self.test:
                frame = self._make_test_frame(time.perf_counter() - t0)
                time.sleep(1 / 240)          # 合成は240fps相当に制限
                ok = True
            else:
                ok, frame = self.cap.read()  # ここがカメラ律速(ブロッキング)
            if not ok:
                self.running = False
                break
            with self.lock:                  # 参照の差し替えのみ(コピー不要)
                self.frame = frame
                self.seq += 1
            # 0.5秒ごとに実測fpsを更新
            n += 1
            now = time.perf_counter()
            if now - win_t0 >= 0.5:
                self.fps = n / (now - win_t0)
                n = 0
                win_t0 = now

    def latest(self):
        """(seq, frame) を返す。フレームが無ければ (0, None)。"""
        with self.lock:
            return self.seq, self.frame

    def stop(self):
        self.running = False
        self.thread.join(timeout=2.0)
        if not self.test:
            self.cap.release()


def main():
    ap = argparse.ArgumentParser(description="camera capture -> OpenGL texture")
    ap.add_argument("--camera", type=int, default=0, help="カメラ番号(既定 0)")
    ap.add_argument("--test", action="store_true", help="カメラ無しの合成映像で実行")
    ap.add_argument("--seconds", type=float, default=0,
                    help="指定秒数で自動終了(0 = ESC まで)")
    args = ap.parse_args()

    cap = CaptureThread(camera_id=args.camera, test=args.test)

    if not glfw.init():
        raise RuntimeError("GLFW の初期化に失敗")
    # 既定の互換プロファイル(macOS では OpenGL 2.1)で十分
    window = glfw.create_window(960, 540, "capture_gl", None, None)
    if not window:
        glfw.terminate()
        raise RuntimeError("ウィンドウ生成に失敗")
    glfw.make_context_current(window)
    glfw.swap_interval(1)   # vsync ON:描画はディスプレイのレートで回る

    tex = glGenTextures(1)
    glBindTexture(GL_TEXTURE_2D, tex)
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR)
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE)
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE)
    glPixelStorei(GL_UNPACK_ALIGNMENT, 1)

    tex_size = None      # 確保済みテクスチャの (w, h)
    last_seq = 0
    disp_n, disp_fps = 0, 0.0
    win_t0 = time.perf_counter()
    start = win_t0

    while not glfw.window_should_close(window):
        if glfw.get_key(window, glfw.KEY_ESCAPE) == glfw.PRESS:
            break
        if args.seconds and time.perf_counter() - start >= args.seconds:
            break

        seq, frame = cap.latest()
        if frame is not None and seq != last_seq:
            last_seq = seq
            h, w = frame.shape[:2]
            if tex_size != (w, h):
                # 初回(または解像度変更時)のみ領域確保
                glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, w, h, 0,
                             GL_BGR, GL_UNSIGNED_BYTE, frame)
                tex_size = (w, h)
            else:
                # 2回目以降は転送のみ(高速)。BGR のまま渡して変換を省く
                glTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, w, h,
                                GL_BGR, GL_UNSIGNED_BYTE, frame)

        fb_w, fb_h = glfw.get_framebuffer_size(window)
        glViewport(0, 0, fb_w, fb_h)
        glClearColor(0.08, 0.08, 0.10, 1.0)
        glClear(GL_COLOR_BUFFER_BIT)

        if tex_size:
            glEnable(GL_TEXTURE_2D)
            glBegin(GL_QUADS)   # 画像は上下反転して貼る(画像は上原点、GLは下原点)
            glTexCoord2f(0, 1); glVertex2f(-1, -1)
            glTexCoord2f(1, 1); glVertex2f(+1, -1)
            glTexCoord2f(1, 0); glVertex2f(+1, +1)
            glTexCoord2f(0, 0); glVertex2f(-1, +1)
            glEnd()
            glDisable(GL_TEXTURE_2D)

        glfw.swap_buffers(window)   # vsync 待ちはここ(描画スレッドだけが待つ)
        glfw.poll_events()

        # 表示側の実測fpsとタイトル更新
        disp_n += 1
        now = time.perf_counter()
        if now - win_t0 >= 0.5:
            disp_fps = disp_n / (now - win_t0)
            disp_n = 0
            win_t0 = now
            glfw.set_window_title(
                window,
                f"capture_gl — camera {cap.fps:5.1f} fps | display {disp_fps:5.1f} fps")

    cap.stop()
    glfw.terminate()


if __name__ == "__main__":
    main()

File: code/capture_gl.cpp
// ============================================================
// capture_gl.cpp — カメラキャプチャ(独立スレッド) + OpenGL テクスチャ描画
//
// キャプチャ専用スレッドがカメラの上限速度でフレームを取り込み、
// メインスレッドは vsync(ディスプレイのリフレッシュレート)に合わせて
// 「最新フレームだけ」をテクスチャとして描画する。
// → 撮影レートが描画レートに律速されない。
//
// ビルド:
//   macOS : brew install opencv glfw pkg-config
//           clang++ -std=c++17 -O2 capture_gl.cpp -o capture_gl \
//             $(pkg-config --cflags --libs opencv4 glfw3) -framework OpenGL
//   Windows: 同梱の CMakeLists.txt + vcpkg を使用(解説ページ参照)
//
// 使い方:
//   ./capture_gl                # カメラ 0 を開く
//   ./capture_gl --camera 1    # カメラ番号を指定
//   ./capture_gl --test        # カメラ無しで合成映像を使う
//   ./capture_gl --seconds 5   # 5秒後に自動終了(動作確認用)
//   ESC または ウィンドウを閉じると終了
//
// タイトルバーに camera fps(取り込み)と display fps(描画)を表示。
// camera fps が display fps を上回っていれば分離が効いている。
// ============================================================

#define GL_SILENCE_DEPRECATION      // macOS の OpenGL 非推奨警告を抑制
#include <GLFW/glfw3.h>
#include <opencv2/opencv.hpp>

#include <atomic>
#include <chrono>
#include <cstdio>
#include <cstring>
#include <mutex>
#include <string>
#include <thread>

using Clock = std::chrono::steady_clock;

static double secondsSince(Clock::time_point t0) {
    return std::chrono::duration<double>(Clock::now() - t0).count();
}

// ---- キャプチャスレッドと共有する状態 --------------------------------
struct SharedFrame {
    std::mutex mtx;
    cv::Mat frame;                    // 最新フレーム(BGR)
    uint64_t seq = 0;                 // フレーム通し番号(新着判定用)
    std::atomic<bool> running{true};
    std::atomic<double> fps{0.0};     // 実測キャプチャfps(0.5秒窓)
};

// カメラ無しでの動作確認用:動くグラデーションと経過時間を描く
static cv::Mat makeTestFrame(int w, int h, double t) {
    cv::Mat img(h, w, CV_8UC3);
    const double phase = t * 2.0;
    for (int y = 0; y < h; ++y) {
        auto* row = img.ptr<cv::Vec3b>(y);
        const double gy = std::cos(y * 6.28 / h + phase) * 0.5 + 0.5;
        for (int x = 0; x < w; ++x) {
            const double gx = std::sin(x * 6.28 / w + phase) * 0.5 + 0.5;
            row[x] = cv::Vec3b(uchar(gx * 255), uchar(gy * 255),
                               uchar((std::sin(phase) * 0.5 + 0.5) * 255));
        }
    }
    char buf[64];
    std::snprintf(buf, sizeof(buf), "TEST %8.3fs", t);
    cv::putText(img, buf, {40, h / 2}, cv::FONT_HERSHEY_SIMPLEX, 2.0,
                {255, 255, 255}, 4);
    return img;
}

// ---- キャプチャスレッド本体 ------------------------------------------
static void captureLoop(SharedFrame& sh, int cameraId, bool test,
                        int width, int height, int reqFps) {
    cv::VideoCapture cap;
    if (!test) {
        cap.open(cameraId);
        if (!cap.isOpened()) {
            std::fprintf(stderr,
                "カメラ %d を開けません。--test で合成映像を試せます\n", cameraId);
            sh.running = false;
            return;
        }
        // 高速化の要:MJPG圧縮・バッファ1枚・解像度/レートの要求
        cap.set(cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M','J','P','G'));
        cap.set(cv::CAP_PROP_FRAME_WIDTH,  width);
        cap.set(cv::CAP_PROP_FRAME_HEIGHT, height);
        cap.set(cv::CAP_PROP_FPS, reqFps);
        cap.set(cv::CAP_PROP_BUFFERSIZE, 1);
    }

    cv::Mat frame;
    int n = 0;
    auto t0 = Clock::now();
    auto winT0 = t0;
    while (sh.running) {
        if (test) {
            frame = makeTestFrame(width, height, secondsSince(t0));
            std::this_thread::sleep_for(std::chrono::microseconds(4166)); // ≈240fps
        } else {
            if (!cap.read(frame)) {      // ここがカメラ律速(ブロッキング)
                sh.running = false;
                break;
            }
        }
        {
            std::lock_guard<std::mutex> lk(sh.mtx);
            cv::swap(sh.frame, frame);   // バッファ交換(コピー無し)
            sh.seq++;
        }
        // 0.5秒ごとに実測fpsを更新
        ++n;
        const double dt = secondsSince(winT0);
        if (dt >= 0.5) {
            sh.fps = n / dt;
            n = 0;
            winT0 = Clock::now();
        }
    }
}

int main(int argc, char** argv) {
    int cameraId = 0;
    bool test = false;
    double seconds = 0;                 // 0 = ESC まで
    for (int i = 1; i < argc; ++i) {
        if (!std::strcmp(argv[i], "--camera") && i + 1 < argc) cameraId = std::atoi(argv[++i]);
        else if (!std::strcmp(argv[i], "--test")) test = true;
        else if (!std::strcmp(argv[i], "--seconds") && i + 1 < argc) seconds = std::atof(argv[++i]);
    }

    SharedFrame sh;
    std::thread capThread(captureLoop, std::ref(sh), cameraId, test, 1280, 720, 120);

    if (!glfwInit()) { sh.running = false; capThread.join(); return 1; }
    // 既定の互換プロファイル(macOS では OpenGL 2.1)で十分
    GLFWwindow* window = glfwCreateWindow(960, 540, "capture_gl", nullptr, nullptr);
    if (!window) { glfwTerminate(); sh.running = false; capThread.join(); return 1; }
    glfwMakeContextCurrent(window);
    glfwSwapInterval(1);                // vsync ON:描画はディスプレイのレートで回る

    GLuint tex = 0;
    glGenTextures(1, &tex);
    glBindTexture(GL_TEXTURE_2D, tex);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
    glPixelStorei(GL_UNPACK_ALIGNMENT, 1);

    int texW = 0, texH = 0;             // 確保済みテクスチャの寸法
    uint64_t lastSeq = 0;
    cv::Mat local;                      // 描画スレッド側のフレーム
    int dispN = 0;
    auto winT0 = Clock::now();
    const auto start = winT0;

    while (!glfwWindowShouldClose(window) && sh.running) {
        if (glfwGetKey(window, GLFW_KEY_ESCAPE) == GLFW_PRESS) break;
        if (seconds > 0 && secondsSince(start) >= seconds) break;

        // 新着フレームがあれば受け取る(ロックは差し替えの瞬間だけ)
        bool fresh = false;
        {
            std::lock_guard<std::mutex> lk(sh.mtx);
            if (sh.seq != lastSeq && !sh.frame.empty()) {
                lastSeq = sh.seq;
                sh.frame.copyTo(local);
                fresh = true;
            }
        }
        if (fresh) {
            if (local.cols != texW || local.rows != texH) {
                // 初回(または解像度変更時)のみ領域確保
                glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, local.cols, local.rows,
                             0, GL_BGR, GL_UNSIGNED_BYTE, local.data);
                texW = local.cols; texH = local.rows;
            } else {
                // 2回目以降は転送のみ(高速)。BGR のまま渡して変換を省く
                glTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, texW, texH,
                                GL_BGR, GL_UNSIGNED_BYTE, local.data);
            }
        }

        int fbW, fbH;
        glfwGetFramebufferSize(window, &fbW, &fbH);
        glViewport(0, 0, fbW, fbH);
        glClearColor(0.08f, 0.08f, 0.10f, 1.0f);
        glClear(GL_COLOR_BUFFER_BIT);

        if (texW > 0) {
            glEnable(GL_TEXTURE_2D);
            glBegin(GL_QUADS);  // 画像は上下反転して貼る(画像は上原点、GLは下原点)
            glTexCoord2f(0, 1); glVertex2f(-1, -1);
            glTexCoord2f(1, 1); glVertex2f(+1, -1);
            glTexCoord2f(1, 0); glVertex2f(+1, +1);
            glTexCoord2f(0, 0); glVertex2f(-1, +1);
            glEnd();
            glDisable(GL_TEXTURE_2D);
        }

        glfwSwapBuffers(window);        // vsync 待ちはここ(描画スレッドだけが待つ)
        glfwPollEvents();

        // 表示側の実測fpsとタイトル更新
        ++dispN;
        const double dt = secondsSince(winT0);
        if (dt >= 0.5) {
            char title[128];
            std::snprintf(title, sizeof(title),
                          "capture_gl — camera %5.1f fps | display %5.1f fps",
                          sh.fps.load(), dispN / dt);
            glfwSetWindowTitle(window, title);
            dispN = 0;
            winT0 = Clock::now();
        }
    }

    sh.running = false;
    capThread.join();
    glfwDestroyWindow(window);
    glfwTerminate();
    return 0;
}
